### PR TITLE
[RISCV][TTI] Add checks for invalid cast operations

### PR DESCRIPTION
**Pull request by arcbbb**
In issue #88802, the LV cost model would query the cost of the TRUNC for source type 2xi1 and destination type 2xi32. This patch adds an early exit check to prevent invalid operations.

### CODE REVIEW

**Comment by arcbbb** [comment]
This fixes a crash in #88802.

**Review by jacquesguan — APPROVED** [review]
LGTM

**Comment by preames** [comment]
If I'm reading the code and discussion right, the LV is querying a truncate with result wider than source?  If so, this does not seem like the right fix.  If LV is querying an invalid argument combination, that's a LV bug and should be fixed as such.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -956,6 +956,9 @@ InstructionCost RISCVTTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
     return getRISCVInstructionCost(Op, DstLT.second, CostKind);
   }
   case ISD::TRUNCATE:
+    // Early return for invalid operation
+    if (Dst->getScalarSizeInBits() >= Src->getScalarSizeInBits())
+      break;
     if (Dst->getScalarSizeInBits() == 1) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:962]: Don't repeatedly call getScalarSizeInBits on the same object. Use a variable.

**Comment by topperc** [comment]
Test case?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Updated. Thanks!

**Comment by arcbbb** [comment]
> Test case?

Added in the updates. Thanks!

**Comment by arcbbb** [comment]
> If I'm reading the code and discussion right, the LV is querying a truncate with result wider than source? If so, this does not seem like the right fix. If LV is querying an invalid argument combination, that's a LV bug and should be fixed as such.

Yes, the actual root cause is LV cost model. I think for TTI cost it is better to return invalid cost or delegate it to the base TTI rather than crash on it,

**Comment by preames** [comment]
> > If I'm reading the code and discussion right, the LV is querying a truncate with result wider than source? If so, this does not seem like the right fix. If LV is querying an invalid argument combination, that's a LV bug and should be fixed as such.
> 
> Yes, the actual root cause is LV cost model. I think for TTI cost it is better to return invalid cost or delegate it to the base TTI rather than crash on it,

TTI should assert sanity of the API, and we should fix the caller.  It's very likely there's a higher level semantics bug in the LV causing this, and we need to find and fix that one.

**Comment by arcbbb** [comment]
> TTI should assert sanity of the API, and we should fix the caller. It's very likely there's a higher level semantics bug in the LV causing this, and we need to find and fix that one.

OK, I create #89161 for this.

**Comment by arcbbb** [comment]
close this since #88802 is no longer RISCV-specific and duplicates #47665.

**Pull request closed without merging**